sklearn/feature_selection/correlation_threshold.py

```diff
@@ -0,0 +1,162 @@
+import numpy as np
+from scipy.sparse import issparse
+
+from ..base import BaseEstimator
+from .base import SelectorMixin
+from ..utils import check_array
+from ..utils.validation import check_is_fitted
+from ..utils.extmath import safe_sparse_dot
+from ..utils.sparsefuncs import min_max_axis, mean_variance_axis
+
+
+class CorrelationThreshold(BaseEstimator, SelectorMixin):
+    """Feature selector that removes features with high correlation.
+
+    This unsupervised feature selection algorithm considers feature pairs with
+    correlation higher than the `threshold` and removes the feature that has
+    the highest mean correlation with the other features. This process is
+    continued till there are no correlations higher than `threshold`.
+
+    Read more in the :ref:`User Guide <correlation_threshold>`.
+
+    Parameters
+    ----------
+    threshold : float, default=0.9
+        Features with a training-set correlation higher than this threshold
+        will be removed.
+
+    Attributes
+    ----------
+    support_mask_ : bool ndarray of shape (n_features, )
+        Boolean mask for features to keep
+
+    Examples
+    --------
+    The following example shows how one of the first highly correlated features
+    are removed, while the uncorrelated feature remains:
+
+    >>> import numpy as np
+    >>> from sklearn.feature_selection import CorrelationThreshold
+    >>> X = np.array([[0.0, 1.0, 2.0], [1.1, 2.0, 3.0], [0.5, 10.1, 1.1]]).T
+    >>> selector = CorrelationThreshold()
+    >>> selector.fit_transform(X)
+    array([[ 1.1,  0.5],
+           [ 2. , 10.1],
+           [ 3. ,  1.1]])
+    """
+    def __init__(self, threshold=0.9):
+        self.threshold = threshold
+
+    def fit(self, X, y=None):
+        """Learn empirical variances from X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} shape of (n_samples, n_features)
+            Training set to compute correlations.
+
+        y : ignored
+            Not used, present here for API consistency by convention.
+
+        Returns
+        -------
+        self : `CorrelationThreshold`
+        """
+        if not (0.0 <= self.threshold <= 1.0):
+            raise ValueError("threshold must be in [0.0, 1.0], got {}".format(
+                             self.threshold))
+
+        X = check_array(X, accept_sparse=['csc', 'csr'], dtype=float)
+
+        n_features = X.shape[1]
+        if self.threshold == 1.0 or (1 in X.shape):
+            self.support_mask_ = np.ones(n_features, dtype=np.bool)
+            return self
+
+        # get constant features
+        if issparse(X):
+            mins, maxes = min_max_axis(X, axis=0)
+            peak_to_peaks = maxes - mins
+            ptp_mask = peak_to_peaks != 0
+
+            # sparse correlation
```

Comment:
Worth putting in utils??

Reply:
In `sparsefuncs`? That seems nice to have.